RavenDB-14515 - added load revisions lazily implementation

omerlewitz opened the pull request.

Commit:
RavenDB-14515 - added load revisions lazily implementation

ayende commented on src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs:
```diff
@@ -64,6 +71,17 @@ public interface IRevisionsSessionOperationsAsync
         /// Returns the number of revisions for specified document.
         /// </summary>
-        Task<long> GetCountForAsync(string id, CancellationToken token = default);
+        Task<long> GetCountForAsync(string id, CancellationToken token = default);//
```
```suggestion
        Task<long> GetCountForAsync(string id, CancellationToken token = default);
```

omerlewitz replied:
done

ayende reviewed (commented):
Looks good, need to also create sync API to wrap it.

Commit:
RavenDB-14515 - added load revisions lazily implementation #12049 => added synchronous lazy implementation

ppekrol commented on src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs:
```diff
@@ -45,18 +44,32 @@ public GetRevisionsCommand(string[] changeVectors, bool metadataOnly = false)
+            return request;
+        }
+
+        public string GetRequestQueryString()
```
I think we can switch it to `private` ?

omerlewitz replied:
done

ppekrol commented on src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs:
```diff
@@ -10,19 +10,25 @@
+    public class DocumentSessionRevisionsAsync : DocumentSessionRevisionsBase, IRevisionsSessionOperationsAsync, ILazyRevisionsOperationsAsync
     {
-        public DocumentSessionRevisionsAsync(InMemoryDocumentSessionOperations session) : base(session)
+        private new AsyncDocumentSession Session { get; }
```
Avoid `new` if possible. Better to cast it inline like you did in sync session.

ppekrol commented on src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs:
```diff
@@ -93,5 +100,43 @@ public async Task<long> GetCountForAsync(string id, CancellationToken token = de
             }
         }
+
+        Lazy<Task<List<T>>> ILazyRevisionsOperationsAsync.GetForAsync<T>(string id, int start = 0, int pageSize = 25, CancellationToken token = default)
```
Can you fix the warnings here?

omerlewitz replied:
done

ppekrol commented on src/Raven.Client/Documents/Session/IRevisionsSessionOperations.cs:
```diff
@@ -74,5 +74,22 @@ public interface IRevisionsSessionOperations
+    /// </summary>
+    public interface ILazyRevisionsOperations
+    {
+        Lazy<List<T>> GetFor<T>(string id, int start = 0, int pageSize = 25);
```
can you copy the comments IRevisionsSessionOperations here? Remember to adjust the text there if needed.

omerlewitz replied:
done

ppekrol commented on src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs:
```diff
@@ -65,5 +72,16 @@ public interface IRevisionsSessionOperationsAsync
+
+    public interface ILazyRevisionsOperationsAsync
+    {
+        Lazy<Task<List<T>>> GetForAsync<T>(string id, int start = 0, int pageSize = 25, CancellationToken token = default);
```
Can you copy the comments from 'IRevisionsSessionOperationsAsync' ?

omerlewitz replied:
done

ppekrol replied:
Ping

omerlewitz replied:
done

ppekrol reviewed (commented):
Looks good.

Commit:
RavenDB-14515 - fixed Pawel's comments

Commit:
RavenDB-14515: fixed - Explicit interface implementation default values,
		       adding comments to ILazyRevisionsOperationsAsync
		       and making sure indentations are good
